src/Parsers/ASTDatabaseOrNone.h

```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include <Parsers/IAST.h>
+
+namespace DB
+{
+
+class ASTDatabaseOrNone : public IAST
+{
+public:
+    bool none = false;
+    String database_name;
```

Comment:
It's not really necessary to have two variables here. I mean if `none` is true then `database_name` is always empty, and on the other hand if `database_name` is empty then `none` must be true. That's why I suggested to use `std::optional<String>` instead of `std::shared_ptr<ASTDatabaseOrNone`, it does the same work but with less amount of code

|    | `std::optional<String>` | `std::shared_ptr<ASTDatabaseOrNone>` |
|----------|:-------------:|------:|
| no DEFAULT DATABASE clause | std::nullopt | nullptr |
| DEFAULT DATABASE `db_name` | "db_name" | {none=false, database_name="db_name"} |
| DEFAULT DATABASE NONE | "" | {none=true, database_name=""} |